### PR TITLE
[BugFix] LazyStackedTensorDict indexing along stack_dim

### PR DESCRIPTION
**Pull request by tcbegley**
## Description

This PR fixes a bug that arose when indexing a `LazyStackedTensorDict` along its stacking dimension.

We detect whether the stacking dimension is included in the dimensions being indexed, index the list of tensordicts, and then apply the remaining indices as necessary. This seemingly was attempted in the original implementation, but there was a logic error with remaining indices not being applied after the list had been indexed.

Previous behaviour

```python
>>> import torch
>>> from tensordict import TensorDict

>>> td = TensorDict({"a": torch.rand(3, 4), "b": torch.rand(3, 4)}, [3, 4])
>>> td = torch.stack([td, td, td], 1)

>>> print(td[2, 2])
TensorDict(
    fields={
        a: Tensor(torch.Size([3, 4]), dtype=torch.float32),
        b: Tensor(torch.Size([3, 4]), dtype=torch.float32)},
    batch_size=torch.Size([3, 4]),
    device=None,
    is_shared=False)
```

New behaviour

```python
>>> print(td[2, 2])
TensorDict(
    fields={
        a: Tensor(torch.Size([4]), dtype=torch.float32),
        b: Tensor(torch.Size([4]), dtype=torch.float32)},
    batch_size=torch.Size([4]),
    device=None,
    is_shared=False)
```

### CODE REVIEW

File: tensordict/tensordict.py [hunk 1]
```diff
@@ -4462,19 +4462,20 @@ def __getitem__(self, item: INDEX_TYPING) -> TensorDictBase:
...
         elif isinstance(item, tuple):
+            # select sub tensordicts
             _sub_item = tuple(
-                _item for i, _item in enumerate(item) if i == self.stack_dim
+                _item for i, _item in enumerate(item) if i != self.stack_dim
             )
-            if len(_sub_item):
-                tensordicts = self.tensordicts[_sub_item[0]]
+
+            if self.stack_dim < len(item):
```
Comment (tcbegley) [inline tensordict/tensordict.py:4470]: this saves us looping over `item` twice.

File: tensordict/tensordict.py [hunk 2]
```diff
@@ -4462,19 +4462,20 @@ def __getitem__(self, item: INDEX_TYPING) -> TensorDictBase:
...
+                tensordicts = self.tensordicts[item[self.stack_dim]]
                 if isinstance(tensordicts, TensorDictBase):
+                    if _sub_item:
+                        return tensordicts[_sub_item]
```
Comment (tcbegley) [inline tensordict/tensordict.py:4474]: this was missing from the original implementation and was the main reason for the bug afaict

File: test/test_tensordict.py [hunk 3]
```diff
@@ -539,6 +539,29 @@ def test_stacked_td(stack_dim, device):
     assert (std5.contiguous() == sub_td.contiguous().unbind(1)[0]).all()
 
 
+@pytest.mark.parametrize("device", get_available_devices())
+@pytest.mark.parametrize("stack_dim", [0, 1])
+def test_stacked_indexing(device, stack_dim):
```
Comment (vmoens) [inline test/test_tensordict.py:544]: One way to do this is to create the stacked td like this

```
tensordict = torch.stack(list(tensordict.unbind(stacked_dim)), stacked_dim)
```
like this you can stack along any dim and all of your tensordicts will have the same shape, hence you can check that the behaviour of indexing is the same in every case

File: test/test_tensordict.py [hunk 4]
```diff
@@ -539,6 +539,29 @@ def test_stacked_td(stack_dim, device):
...
+        for _ in range(3)
+    ]
+    tds = torch.stack(tensordicts, stack_dim)
+
+    item = (2, 2)
+    assert tds[item].batch_size == torch.Size([5])
+    assert (tds[item].get("a") == tds.get("a")[item]).all()
+    assert (tds[item].get("a") == tensordicts[2].get("a")[2]).all()
+
+    item = (slice(1, 2), 2)
```
Comment (vmoens) [inline test/test_tensordict.py:560]: can we test with ellipsis too?

**Review by vmoens — APPROVED** [review]
LGTM, some minor comments

File: test/test_tensordict.py (hunk 4 above)
Comment (tcbegley) [reply test/test_tensordict.py]: Yep, added

File: test/test_tensordict.py (hunk 3 above)
Comment (tcbegley) [reply test/test_tensordict.py:544]: Neat, I've updated the test and used this pattern.

**Review by vmoens — APPROVED** [review]
Wonderful
